lib/source-destination/url-compressed-source.ts

```diff
@@ -0,0 +1,386 @@
+		);
+	}
+
+	private findImagePart(imageJSON: ImageJSON, image: string): ImageJSONPart {
+		const [part] = imageJSON[image]?.parts ?? [];
+		if (part === undefined) {
+			throw new Error(`Couldn't find compressed part for image ${image}`);
+		}
+		return part;
+	}
+
+	private findPart(definition: FileOnPartition): ImageJSONPart {
+		if (definition.partition !== undefined) {
+			const partition = normalizePartition(definition.partition);
+			return this.findPartitionPart(this.imageJSON, partition);
+		} else if (definition.image !== undefined) {
+			return this.findImagePart(this.imageJSON, definition.image);
+		} else {
+			throw new Error('No partition or image to configure found');
+		}
+	}
+
+	private async extractDeflateToDisk(filename: string) {
+		const stream = await this.getPartStream(filename);
+		const combined = CombinedStream.create();
+		combined.append(stream);
+		combined.append(DEFLATE_END);
+		const inflate = createInflateRaw();
+		combined.pipe(inflate);
+		return new BufferDisk(await streamToBuffer(inflate));
+	}
+
+	private async configure() {
+		if (this.configuration === undefined) {
+			return;
+		}
+		const disks = new Map<string, BufferDisk>();
+		const self = this;
+		async function getDisk(definition: FileOnPartition): Promise<BufferDisk> {
+			const filename = self.findPart(definition).filename;
+			const d = disks.get(filename);
+			if (d !== undefined) {
+				return d;
+			}
+			const d2 = await self.extractDeflateToDisk(filename);
+			disks.set(filename, d2);
+			return d2;
+		}
+		// configure
+		if (this.deviceTypeJSON?.configuration?.config) {
+			await configure(
+				await getDisk(this.deviceTypeJSON.configuration.config),
+				undefined,
+				this.configuration,
+			);
+		}
+		// copy operations
+		for (const cpy of this.deviceTypeJSON?.configuration?.operations ?? []) {
+			if (shouldRunOperation(this.configuration ?? {}, cpy)) {
+				await copy(
+					await getDisk(cpy.from),
+					undefined,
+					cpy.from.path,
+					await getDisk(cpy.to),
+					undefined,
+					cpy.to.path,
+				);
+			}
+		}
+		// compress
+		await Promise.all(
+			Array.from(disks.entries()).map(async ([filename, disk]) => {
+				const stream = (await disk.getStream()).pipe(createDeflatePart());
+				const buffer = await streamToBuffer(stream);
+				const { crc, zLen } = stream.metadata();
+				this.configuredParts.set(filename, { crc, zLen, buffer });
+			}),
+		);
+	}
+
+	protected async _open(): Promise<void> {
+		// Validate that all required URLs are present
+		if (!this.urls.VERSION || !this.urls['image.json']) {
+			throw new Error('Required URLs (VERSION, image.json) must be provided');
+		}
+
+		const [
+			{ supervisorVersion, lastModified },
+			osVersion,
+			imageJSON,
+			deviceTypeJSON,
+		] = await Promise.all([
+			this.getSupervisorVersion(),
+			this.getOsVersion(),
+			this.getImageJSON(),
+			this.getDeviceTypeJSON(),
+		]);
+		if (deviceTypeJSON?.yocto?.archive) {
```

Comment:
I think we should trust the `getDeviceTypeJSON` response type and since that's a critical part perhaps throw from within `getDeviceTypeJSON` if we can't retrieve it.
```suggestion
		if (deviceTypeJSON.yocto?.archive) {
```
also, do we actually need the `?` in the  `yocto?.archive` part? Afaict `balena-s3-compressed-source.ts` doesn't have such a `?.` .